lib/spack/spack/config.py

```diff
@@ -763,57 +764,93 @@ def _merge_yaml(dest, source):
+
+            # ensure that keys are marked in the destination. The
+            # key_marks dict ensures we can get the actual source key
+            # objects from dest keys
+            for dk in list(dest.keys()):
+                if dk in key_marks and syaml.markable(dk):
+                    syaml.mark(dk, key_marks[dk])
+                elif dk in key_marks:
+                    # The destination key may not be markable if it was derived
+                    # from a schema default. In this case replace the key.
+                    val = dest.pop(dk)
+                    dest[key_marks[dk]] = val
+
+            return dest
+
+    # If we reach here, overwrite with a copy of the source value.
+    return copy.copy(source)
 
-    # Source list is prepended (for precedence)
-    if they_are(list):
-        dest[:] = source + [x for x in dest if x not in source]
-        return dest
 
-    # Source dict is merged into dest.
-    elif they_are(dict):
-        # track keys for marking
-        key_marks = {}
+#
+# Process a path argument to config.set() that may contain overrides ('::' or
+# trailing ':')
+#
+def _process_path(path):
+    result = []
+    if path.startswith(':'):
+        raise syaml.SpackYAMLError("Illegal leading `:' in path `{0}'".
+                                   format(path), '')
+    while path:
+        front, _, path = path.partition(':')
+        if (_ and not path) or path.startswith(':'):
+            path = path.lstrip(':')
+            front = syaml.syaml_str(front)
+            front.override = True
+        result.append(front)
+    return result
 
-        for sk, sv in iteritems(source):
-            if _override(sk) or sk not in dest:
-                # if sk ended with ::, or if it's new, completely override
-                dest[sk] = copy.copy(sv)
-            else:
-                # otherwise, merge the YAML
-                dest[sk] = _merge_yaml(dest[sk], source[sk])
-
-            # this seems unintuitive, but see below. We need this because
-            # Python dicts do not overwrite keys on insert, and we want
-            # to copy mark information on source keys to dest.
-            key_marks[sk] = sk
-
-        # ensure that keys are marked in the destination.  the key_marks dict
-        # ensures we can get the actual source key objects from dest keys
-        for dk in list(dest.keys()):
-            if dk in key_marks and syaml.markable(dk):
-                syaml.mark(dk, key_marks[dk])
-            elif dk in key_marks:
-                # The destination key may not be markable if it was derived
-                # from a schema default. In this case replace the key.
-                val = dest.pop(dk)
-                dest[key_marks[dk]] = val
-
-        return dest
-
-    # In any other case, overwrite with a copy of the source value.
-    else:
-        return copy.copy(source)
+
+#
+# Process a dict whose keys may describe overrides (trailing ':')
+#
+def _process_dict(data):
```

Comment:
Could you choose a more-descriptive name for this function and mention that it is a helper function for `InternalConfigScope` (since generally there are other mechanisms to handle overrides on other config objects)?